doc/design/custom_xcode_schemes.md

```diff
@@ -0,0 +1,467 @@
+# Custom Xcode Schemes
+
+This document is a proposal for how custom Xcode schemes can be defined and implemented in
+`rules_xcodeproj`.
+
+## Contents
+
+* [Automatic Scheme Generation](#automatic-scheme-generation)
+* [Introduction of Custom Xcode Schemes](#introduction-of-custom-xcode-schemes)
+  * [Defining a Simple Xcode Scheme](#defining-a-simple-xcode-scheme)
+  * [Specifying a Launch Target](#specifying-a-launch-target)
+  * [Specifying Launch Arguments and Environment Variables](#specifying-launch-arguments-and-environment-variables)
+* [Introduction of Scheme Autogeneration Mode](#introduction-of-scheme-autogeneration-mode)
+* [Implementation Changes](#implementation-changes)
+  * [xcode\_schemes Module](#xcode_schemes-module)
+  * [Changes to \_xcodeproj Rule](#changes-to-_xcodeproj-rule)
+    * [scheme\_autogeneration\_mode Attribute](#scheme_autogeneration_mode-attribute)
+    * [schemes Attribute](#schemes-attribute)
+  * [Changes to xcodeproj Macro](#changes-to-xcodeproj-macro)
+    * [Collecting Targets from Schemes](#collecting-targets-from-schemes)
+* [Configuration and Target Selection in Schemes](#configuration-and-target-selection-in-schemes)
+* [Outstanding Questions](#outstanding-questions)
+
+## Automatic Scheme Generation
+
+As of this writing, the `rules_xcodeproj` ruleset generates an Xcode scheme for every buildable
+target provided to the `xcodeproj` rule. This allows a client to quickly define an `xcodeproj`
+target and generate an Xcode project.
+
+Let's start with an example.
+
+```python
+# Assumptions
+#   //Sources/Foo:Foo - swift_library
+#   //Sources/FooTests:FooTests = ios_unit_test
+
+xcodeproj(
+    name = "generate_xcodeproj",
+    project_name = "Command Line",
+    tags = ["manual"],
+    targets = [
+        "//Sources/Foo",
```

Comment:
Unless we want to get into the intricacies of how this will actually create _3 targets_ and _3 schemes_ let's not include the library in `targets`.

```suggestion
#   //Sources/FooTests:FooTests = ios_unit_test
#   //Sources/Foo:Foo - swift_library, a dependency of //Sources/FooTests:FooTests

xcodeproj(
    name = "generate_xcodeproj",
    project_name = "Command Line",
    tags = ["manual"],
    targets = [
```